Make cyborgs magnetize items

GR1231 opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
Makes cyborgs magnetize ore with integrated ore bags.

## Why / Balance
Fixes #39906
Saves the time and annoyance of manually clicking to pickup ores when playing as a salvage cyborg.

## Technical details
<!-- Summary of code changes for easier review. -->
Modifies the MagnetPickupSystem.
Adds an if statement that checks if the SlotFlags datafield is null.
If it is, the checks for Inventory SlotFlags (if the ore bags are in belt slots) get skipped.

Adds a new integrated bag of holding item and adds it to the advanced mining module.

Currently cyborgs always magnetize ore, even when the module is not selected.

## Media
https://github.com/user-attachments/assets/b289b936-a16c-43c1-8697-babe14dc98ff



## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Note
I am still learning the codebase and the conventions and coding like this in general so if I did something horrible or unwanted please let me know.

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
:cl:
- add: Salvage cyborgs can now magnetize ore with integrated ore bags.

GR1231 committed:
Make borgs magnetize ore

MilenVolf commented on Content.Shared/Storage/EntitySystems/MagnetPickupSystem.cs:
```diff
@@ -48,17 +52,21 @@ public override void Update(float frameTime)
 
-            if ((slotDef.SlotFlags & comp.SlotFlags) == 0x0)
-                continue;
+            if (!_entityManager.TryGetComponent<BorgSwitchableTypeComponent>(parentUid, out _) || !_hands.TryGetActiveItem(parentUid, out var activeItem) && activeItem != uid)
```
If you don't need "out" variable, use ```HasComp<AnyComponent>(uid)``` instead.

```suggestion
            if (!HasComp<BorgSwitchableTypeComponent>(parentUid) || !_hands.TryGetActiveItem(parentUid, out var activeItem) && activeItem != uid)
```

After this don't forget to remove EntityManager from dependency list (it'll be not used in this case)

MilenVolf reviewed (changes requested):
Good PR! Just a minor nitpick.

> One small issue is that the bag magnetizes when the module is selected not the item itself, probably due to how borg hands work.

I actually prefer the current behaviour. It's more user-friendly for borgs, so I don't think it needs to be changed.

> I am still learning the codebase and the conventions and coding like this in general so if I did something horrible or unwanted please let me know.

You're doing great, keep it up 👍

iaada reviewed (changes requested):
I'm not a fan of hard coding borgs as an exception into this system. What I'd like to see instead is to make `SlotFlags` nullable in the component, and to skip the slot check if the datafield is null. Then change the borg bag in yml.

GR1231 commented:
> I'm not a fan of hard coding borgs as an exception into this system. What I'd like to see instead is to make `SlotFlags` nullable in the component, and to skip the slot check if the datafield is null. Then change the borg bag in yml.

Got it, will try to do it that way instead.

GR1231 committed:
nullable SlotFlags

GR1231 commented:
Changed it so there is only the null check in the magnet system. This current implementation assumes that only borgs can access the integrated bags. Also added a new integrated ore bag of holding item for the upgraded module.

Currently borgs always pickup ore around them even if a different module is selected but that can be "fixed" with the active hand check next to the new null check, but I removed it for now.

iaada reviewed (commented):
I think you should also add the active hand check, because it activating when the module is closed doesn't feel right.

GR1231 committed:
add hand check

iaada commented on Content.Shared/Storage/EntitySystems/MagnetPickupSystem.cs:
```diff
@@ -48,17 +50,21 @@ public override void Update(float frameTime)
 
-            if ((slotDef.SlotFlags & comp.SlotFlags) == 0x0)
-                continue;
+            if (comp.SlotFlags != null || !_hands.TryGetActiveItem(parentUid, out var activeItem) && activeItem != uid)
```
This should be its own bool for more control. You're right side logic here also always evaluates false since `!_hands.TryGetActiveItem()` will be false when it finds an item, but `activeItem != uid` will be false if there's nothing in hand.

```suggestion
            if (comp.RequireActiveHand && (!_hands.TryGetActiveItem(parentUid, out var activeItem) || activeItem != uid))
                continue;
                
            if (comp.SlotFlags != null)
```

iaada commented on Content.Shared/Storage/Components/MagnetPickupComponent.cs:
```diff
@@ -16,7 +16,7 @@ public sealed partial class MagnetPickupComponent : Component
     /// What container slot the magnet needs to be in to work.
     /// </summary>
     [ViewVariables(VVAccess.ReadWrite), DataField("slotFlags")]
```
This is an old syntax. `ViewVariables(VVAccess.ReadWrite)` is now included as a part of `DataField`. `DataField` also now automatically creates an alias in camelcase. Since you're here, could you also please clean up this component a lil?

```suggestion
    [DataField]
```

GR1231 committed:
fix and cleanup

GR1231 commented:
I did the changes and cleanup you asked, sorry for my sloppy code and logic.
Because the hand checks were wrong the bags magnetized when the module was selected, but now it only sucks ore when the item itself is selected (the active hand). Dont know if that is acceptable behavior but I like it.

iaada reviewed (approved):
It works and looks good to me! Thank you for the clean up as well. One last thing: Can you edit the description of the borg bags to explain that it only magnetizes when held? That way way it's a bit easier to discover the functionality.

GR1231 committed:
description edit

SlamBamActionman reviewed (approved):
Looks good to me! Thank you for your contribution.
